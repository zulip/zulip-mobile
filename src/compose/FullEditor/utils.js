/* @flow */
import regexValidator from './webLinkValidator';

export const replaceBetween = (text: string, selection: Object, what: string) =>
  text.substring(0, selection.start) + what + text.substring(selection.end);

export const isStringWebLink = (text: string): boolean => {
  const pattern = regexValidator;
  return pattern.test(text);
};
