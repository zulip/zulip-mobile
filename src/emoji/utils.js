/* @flow */
import type { RealmEmojiType, ZulipExtraEmojiType } from '../types';
import { codePointMap } from './emojiMap';

export const getRealmEmojiWithName = (realmEmoji: Object, emojiName: string): RealmEmojiType =>
  realmEmoji[Object.keys(realmEmoji).find(key => realmEmoji[key].name === emojiName)];

export const getZulipExtraEmojiWithName = (
  zulipExtraEmojis: Object,
  emojiName: string,
): ZulipExtraEmojiType => zulipExtraEmojis[emojiName];

export const getEmojiTypeFromName = (
  realmEmoji: Object,
  zulipExtraEmojis: Object,
  emojiName: string,
): string | null => {
  if (getZulipExtraEmojiWithName(zulipExtraEmojis, emojiName)) {
    return 'zulip_extra_emoji';
  } else if (getRealmEmojiWithName(realmEmoji, emojiName)) {
    return 'realm_emoji';
  } else if (codePointMap[emojiName]) {
    return 'unicode_emoji';
  }
  return null;
};

export const getEmojiCodeFromName = (
  realmEmoji: Object,
  zulipExtraEmojis: Object,
  emojiName: string,
): string | null => {
  if (getZulipExtraEmojiWithName(zulipExtraEmojis, emojiName)) {
    return emojiName;
  }
  const emoji = getRealmEmojiWithName(realmEmoji, emojiName);
  if (emoji) {
    return emoji.id.toString();
  } else if (codePointMap[emojiName]) {
    return codePointMap[emojiName];
  }

  return null;
};
