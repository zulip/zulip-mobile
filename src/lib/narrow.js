export const homeNarrow = {};

export const mentionedNarrow = [{
  negated: false,
  operator: 'is',
  operand: 'mentioned',
}];

export const privateNarrow = (emails) => {
  if (!emails) {
    return [{
      negated: false,
      operator: 'is',
      operand: 'private',
    }];
  }
  return [{
    negated: false,
    operator: 'pm-with',
    operand: emails.join(),
  }];
};

export const starredNarrow = [{
  negated: false,
  operator: 'is',
  operand: 'starred',
}];

export const streamNarrow = (stream) => [{
  negated: false,
  operator: 'stream',
  operand: stream,
}];

export const topicNarrow = (stream, topic) => [
  {
    negated: false,
    operator: 'stream',
    operand: stream,
  },
  {
    negated: false,
    operator: 'topic',
    operand: topic,
  },
];
