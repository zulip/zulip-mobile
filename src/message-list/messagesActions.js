import { getMessages } from '../api';
import {
  MESSAGE_FETCH_START,
  MESSAGE_FETCH_SUCCESS,
} from '../constants';

export const fetchMessages = (
  auth,
  anchor: number,
  numBefore: number,
  numAfter: number,
  narrow,
) =>
  async (dispatch) => {
    dispatch({ type: MESSAGE_FETCH_START, narrow });

    const messages = await getMessages(auth, anchor, numBefore, numAfter, narrow);

    dispatch({
      type: MESSAGE_FETCH_SUCCESS,
      auth,
      messages,
      anchor,
      narrow,
      caughtUp: (numAfter >= numBefore &&
        messages.length < numAfter + numBefore + (numBefore ? 1 : 0)),
    });
  };
