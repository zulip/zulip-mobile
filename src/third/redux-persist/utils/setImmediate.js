const hasNativeSupport = typeof global !== 'undefined' && typeof global.setImmediate !== 'undefined'
const setImmediate = hasNativeSupport ? (fn, ms) => global.setImmediate(fn, ms) : (fn, ms) => setTimeout(fn, ms)

export default setImmediate
