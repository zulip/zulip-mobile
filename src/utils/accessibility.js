/* @flow */
export const textWithUnreadCount = (text: string, unreadCount: number = 0): string => {
  if (unreadCount === 0) {
    return text;
  }

  return `${text}, ${unreadCount} unread message${unreadCount > 1 ? 's' : ''}`;
};
