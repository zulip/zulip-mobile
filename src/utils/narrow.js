/* @flow strict-local */
import isEqual from 'lodash.isequal';
import unescape from 'lodash.unescape';

import type { Narrow, Message, Outbox } from '../types';
import { normalizeRecipients } from './recipient';

export const isSameNarrow = (narrow1: Narrow, narrow2: Narrow): boolean =>
  Array.isArray(narrow1) && Array.isArray(narrow2) && isEqual(narrow1, narrow2);

export const parseNarrowString = (narrowStr: string): Narrow => JSON.parse(unescape(narrowStr));

export const HOME_NARROW: Narrow = [];

export const HOME_NARROW_STR: string = '[]';

export const privateNarrow = (email: string): Narrow => [
  {
    operator: 'pm-with',
    operand: email,
  },
];

export const groupNarrow = (emails: string[]): Narrow => [
  {
    operator: 'pm-with',
    operand: emails.join(),
  },
];

export const specialNarrow = (operand: string): Narrow => [
  {
    operator: 'is',
    operand,
  },
];

export const STARRED_NARROW = specialNarrow('starred');

export const STARRED_NARROW_STR = JSON.stringify(STARRED_NARROW);

export const MENTIONED_NARROW = specialNarrow('mentioned');

export const MENTIONED_NARROW_STR = JSON.stringify(MENTIONED_NARROW);

export const ALL_PRIVATE_NARROW = specialNarrow('private');

export const ALL_PRIVATE_NARROW_STR = JSON.stringify(ALL_PRIVATE_NARROW);

export const streamNarrow = (stream: string): Narrow => [
  {
    operator: 'stream',
    operand: stream,
  },
];

export const topicNarrow = (stream: string, topic: string): Narrow => [
  {
    operator: 'stream',
    operand: stream,
  },
  {
    operator: 'topic',
    operand: topic,
  },
];

export const SEARCH_NARROW = (query: string): Narrow => [
  {
    operator: 'search',
    operand: query,
  },
];

type NarrowCases<T> = {|
  home: () => T,
  pm: (email: string) => T,
  groupPm: (emails: string[]) => T,
  starred: () => T,
  mentioned: () => T,
  allPrivate: () => T,
  stream: (name: string) => T,
  topic: (streamName: string, topic: string) => T,
  search: (query: string) => T,
|};

/* prettier-ignore */
export function caseNarrow<T>(narrow: Narrow, cases: NarrowCases<T>): T {
  const err = (): empty => {
    throw new Error(`bad narrow: ${JSON.stringify(narrow)}`);
  };

  switch (narrow.length) {
    case 0: return cases.home();
    case 1:
      switch (narrow[0].operator) {
        case 'pm-with':
          if (narrow[0].operand.indexOf(',') < 0) {
            return cases.pm(narrow[0].operand);
          } else { /* eslint-disable-line */
            const emails = narrow[0].operand.split(',');
            return cases.groupPm(emails);
          }
        case 'is':
          switch (narrow[0].operand) {
            case 'starred': return cases.starred();
            case 'mentioned': return cases.mentioned();
            case 'private': return cases.allPrivate();
            default: return err();
          }
        case 'stream': return cases.stream(narrow[0].operand);
        case 'search': return cases.search(narrow[0].operand);
        default: return err();
      }
    case 2: return cases.topic(narrow[0].operand, narrow[1].operand);
    default: return err();
  }
}

export function caseNarrowPartial<T>(narrow: Narrow, cases: $Shape<NarrowCases<T>>): T {
  const err = (type: string) => (): empty => {
    throw new Error(`unexpected ${type} narrow: ${JSON.stringify(narrow)}`);
  };
  return caseNarrow(
    narrow,
    Object.assign(
      ({
        home: err('home'),
        pm: err('PM'),
        groupPm: err('group PM'),
        starred: err('starred'),
        mentioned: err('mentions'),
        allPrivate: err('all-private'),
        stream: err('stream'),
        topic: err('topic'),
        search: err('search'),
      }: NarrowCases<T>),
      cases,
    ),
  );
}

export function caseNarrowDefault<T>(
  narrow: Narrow,
  cases: $Shape<NarrowCases<T>>,
  defaultCase: () => T,
): T {
  return caseNarrow(
    narrow,
    Object.assign(
      ({
        home: defaultCase,
        pm: defaultCase,
        groupPm: defaultCase,
        starred: defaultCase,
        mentioned: defaultCase,
        allPrivate: defaultCase,
        stream: defaultCase,
        topic: defaultCase,
        search: defaultCase,
      }: NarrowCases<T>),
      cases,
    ),
  );
}

export const isHomeNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { home: () => true }, () => false);

export const isPrivateNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { pm: () => true }, () => false);

export const isGroupNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { groupPm: () => true }, () => false);

/**
 * The recipients' emails if a group PM narrow; else error.
 *
 * Any use of this probably means something higher up should be refactored
 * to use caseNarrow.
 */
export const emailsOfGroupNarrow = (narrow: Narrow): string[] =>
  caseNarrowPartial(narrow, { groupPm: emails => emails });

/**
 * The recipient's email if a Private narrow; else error.
 */
export const emailOfPrivateNarrow = (narrow: Narrow): string =>
  caseNarrowPartial(narrow, { pm: email => email });

export const isPrivateOrGroupNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { pm: () => true, groupPm: () => true }, () => false);

export const isSpecialNarrow = (narrow?: Narrow): boolean =>
  !!narrow
  && caseNarrowDefault(
    narrow,
    { starred: () => true, mentioned: () => true, allPrivate: () => true },
    () => false,
  );

export const isAllPrivateNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { allPrivate: () => true }, () => false);

export const isStreamNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { stream: () => true }, () => false);

export const isTopicNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { topic: () => true }, () => false);

export const isStreamOrTopicNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { stream: () => true, topic: () => true }, () => false);

export const isSearchNarrow = (narrow?: Narrow): boolean =>
  !!narrow && caseNarrowDefault(narrow, { search: () => true }, () => false);

/** (For search narrows, just returns false.) */
export const isMessageInNarrow = (message: Message, narrow: Narrow, ownEmail: string): boolean => {
  const matchRecipients = (emails: string[]) => {
    const normalizedRecipients = normalizeRecipients(message.display_recipient);
    const normalizedNarrow = [...emails, ownEmail].sort().join(',');
    return normalizedRecipients === ownEmail || normalizedRecipients === normalizedNarrow;
  };

  const { flags } = message;
  if (!flags) {
    throw new Error('`message.flags` should be defined.');
  }

  return caseNarrow(narrow, {
    home: () => true,
    stream: name => name === message.display_recipient,
    topic: (streamName, topic) =>
      streamName === message.display_recipient && topic === message.subject,
    pm: email => matchRecipients([email]),
    groupPm: matchRecipients,
    starred: () => flags.includes('starred'),
    mentioned: () => flags.includes('mentioned') || flags.includes('wildcard_mentioned'),
    allPrivate: () => message.type === 'private',
    search: () => false,
  });
};

export const canSendToNarrow = (narrow: Narrow): boolean =>
  caseNarrow(narrow, {
    pm: () => true,
    groupPm: () => true,
    stream: () => true,
    topic: () => true,
    home: () => false,
    starred: () => false,
    mentioned: () => false,
    allPrivate: () => false,
    search: () => false,
  });

/** True just if `haystack` contains all possible messages in `needle`. */
export const narrowContains = (haystack: Narrow, needle: Narrow): boolean => {
  if (isHomeNarrow(haystack)) {
    return true;
  }
  if (isAllPrivateNarrow(haystack) && isPrivateOrGroupNarrow(needle)) {
    return true;
  }
  if (isStreamNarrow(haystack) && needle[0].operand === haystack[0].operand) {
    return true;
  }
  return JSON.stringify(needle) === JSON.stringify(haystack);
};

export const getNarrowFromMessage = (message: Message | Outbox, ownEmail: string) => {
  if (Array.isArray(message.display_recipient)) {
    const recipient =
      message.display_recipient.length > 1
        ? message.display_recipient.filter(x => x.email !== ownEmail)
        : message.display_recipient;
    return groupNarrow(recipient.map(x => x.email));
  }

  if (message.subject && message.subject.length) {
    return topicNarrow(message.display_recipient, message.subject);
  }

  return streamNarrow(message.display_recipient);
};
