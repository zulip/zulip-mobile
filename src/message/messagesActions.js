/* @flow */
import type { Narrow, Dispatch, GetState, GlobalState, SwitchNarrowAction } from '../types';
import config from '../config';
import { NULL_ARRAY, NULL_CAUGHTUP } from '../nullObjects';
import { getAuth, getUsers, getAllMessages, getStreams, getIsHydrated } from '../selectors';
import { SWITCH_NARROW } from '../actionConstants';
import { getMessageIdFromLink, getNarrowFromLink, isUrlInAppLink, getFullUrl } from '../utils/url';
import openLink from '../utils/openLink';
import { fetchMessagesAtFirstUnread, fetchMessagesAroundAnchor } from './fetchActions';
import { validateNarrow, fixStreamNarrowOperand } from '../utils/narrow';

export const switchNarrow = (narrow: Narrow): SwitchNarrowAction => ({
  type: SWITCH_NARROW,
  narrow,
});

const isNarrowValid = (narrow: Narrow, state: GlobalState): boolean =>
  validateNarrow(narrow, getStreams(state), getUsers(state));

const fixStreamNarrow = (narrow: Narrow, state: GlobalState) =>
  fixStreamNarrowOperand(narrow, getStreams(state));

export const doNarrow = (narrow: Narrow, anchor: number = 0) => (
  dispatch: Dispatch,
  getState: GetState,
) => {
  const state = getState();

  fixStreamNarrow(narrow, state);

  if (!isNarrowValid(narrow, state) || !getIsHydrated(state)) {
    return;
  }

  dispatch(switchNarrow(narrow));

  const allMessages = getAllMessages(state);
  const messagesForNarrow = allMessages[JSON.stringify(narrow)] || NULL_ARRAY;
  const tooFewMessages = messagesForNarrow.length < config.messagesPerRequest / 2;

  const caughtUp = state.caughtUp[JSON.stringify(narrow)] || NULL_CAUGHTUP;
  const isCaughtUp = caughtUp.newer && caughtUp.older;

  if (anchor === 0 && tooFewMessages && !isCaughtUp) {
    dispatch(fetchMessagesAtFirstUnread(narrow));
  } else if (anchor !== 0) {
    dispatch(fetchMessagesAroundAnchor(narrow, anchor));
  }
};

export const messageLinkPress = (href: string) => (dispatch: Dispatch, getState: GetState) => {
  const state = getState();
  const auth = getAuth(state);

  if (isUrlInAppLink(href, auth.realm)) {
    const users = getUsers(state);
    const anchor = getMessageIdFromLink(href, auth.realm);
    const narrow = getNarrowFromLink(href, auth.realm, users);

    dispatch(doNarrow(narrow, anchor));
  } else {
    openLink(getFullUrl(href, auth.realm));
  }
};
