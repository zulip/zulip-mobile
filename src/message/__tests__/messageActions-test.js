import mockStore from 'redux-mock-store'; // eslint-disable-line

import { doNarrow } from '../messagesActions';
import { streamNarrow, homeNarrow } from '../../utils/narrow';

const narrow = streamNarrow('some stream');
const streamNarrowStr = JSON.stringify(narrow);

describe('messageActions', () => {
  describe('doNarrow', () => {
    test('when no messages in new narrow and caughtUp is false, actions to fetch messages and switch narrow are dispatched', () => {
      const store = mockStore({
        caughtUp: {
          [streamNarrowStr]: {
            newer: false,
            older: true,
          },
        },
        chat: {
          narrow: homeNarrow,
          messages: {},
        },
      });

      store.dispatch(doNarrow(narrow));
      expect(store.getActions()).toMatchSnapshot();
    });

    test('when no messages in new narrow and caughtUp is true, only action to switch narrow is dispatched', () => {
      const store = mockStore({
        caughtUp: {
          [streamNarrowStr]: {
            newer: true,
            older: true,
          },
        },
        chat: {
          narrow: homeNarrow,
          messages: {},
        },
      });

      store.dispatch(doNarrow(narrow));
      expect(store.getActions()).toMatchSnapshot();
    });

    test('when messages in new narrow, only action to switch narrow is dispatched', () => {
      const store = mockStore({
        caughtUp: {},
        chat: {
          narrow: homeNarrow,
          messages: {
            [streamNarrowStr]: [{ id: 1 }],
          },
        },
      });

      store.dispatch(doNarrow(narrow));
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
