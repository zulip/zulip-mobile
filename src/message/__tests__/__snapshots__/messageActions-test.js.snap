// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`messageActions doNarrow when messages in new narrow, only action to switch narrow is dispatched 1`] = `
Array [
  Object {
    "narrow": Array [
      Object {
        "operand": "some stream",
        "operator": "stream",
      },
    ],
    "type": "SWITCH_NARROW",
  },
]
`;

exports[`messageActions doNarrow when no messages in new narrow and caughtUp is false, actions to fetch messages and switch narrow are dispatched 1`] = `
Array [
  Object {
    "narrow": Array [
      Object {
        "operand": "some stream",
        "operator": "stream",
      },
    ],
    "type": "SWITCH_NARROW",
  },
  Object {
    "narrow": Array [
      Object {
        "operand": "some stream",
        "operator": "stream",
      },
    ],
    "numAfter": 15,
    "numBefore": 5,
    "type": "MESSAGE_FETCH_START",
  },
]
`;

exports[`messageActions doNarrow when no messages in new narrow and caughtUp is true, only action to switch narrow is dispatched 1`] = `
Array [
  Object {
    "narrow": Array [
      Object {
        "operand": "some stream",
        "operator": "stream",
      },
    ],
    "type": "SWITCH_NARROW",
  },
  Object {
    "narrow": Array [
      Object {
        "operand": "some stream",
        "operator": "stream",
      },
    ],
    "numAfter": 15,
    "numBefore": 5,
    "type": "MESSAGE_FETCH_START",
  },
]
`;
